btsensor/src/bthome/v2.rs

```diff
@@ -0,0 +1,493 @@
+}
+
+impl BtHomeV2 {
+    /// Attempts to decode the given service data as a BTHome v2 advertisement.
+    pub fn decode(data: &[u8]) -> Result<Self, DecodeError> {
+        let device_info = data[0];
+        let encrypted = device_info & DEVICE_INFO_ENCRYPTED != 0;
+        let trigger_based = device_info & DEVICE_INFO_TRIGGER_BASED != 0;
+        let version = (device_info & DEVICE_INFO_VERSION_MASK) >> DEVICE_INFO_VERSION_OFFSET;
+        if version != 2 {
+            return Err(DecodeError::UnsupportedVersion(version));
+        }
+
+        let mut remaining_data = &data[1..];
+        let mut elements = Vec::new();
+        while remaining_data.len() >= 2 {
+            let (element, element_length) = Element::decode(remaining_data)?;
+            remaining_data = &remaining_data[element_length..];
+            elements.push(element);
+        }
+
+        Ok(Self {
+            encrypted,
+            trigger_based,
+            elements,
+        })
+    }
+}
+
+macro_rules! generate_element {
+    [$({$object_id:literal, $name:ident, $type:ty, $reader:ident, $display_name:literal, $unit:literal},)*] => {
+        /// A single element of a BTHome v2 advertisement: either a sensor reading or an event.
+        #[derive(Clone, Debug, Eq, PartialEq)]
+        pub enum Element {
+            $( $name($type), )*
+        }
+
+        impl Element {
+            /// Returns the name of the property.
+            pub fn name(&self) -> &'static str {
+                match self {
+                    $( Self::$name(_) => $display_name, )*
+                }
+            }
+
+            /// Returns the standard unit for the property.
+            pub fn unit(&self) -> &'static str {
+                match self {
+                    $( Self::$name(_) => $unit, )*
+                }
+            }
+
+            fn decode(data: &[u8]) -> Result<(Self, usize), DecodeError> {
+                let object_id = data[0];
+                let remaining = &data[1..];
+                match object_id {
+                    0x0a => {
+                        let (value, length) = read_u24(remaining)?;
+                        Ok((Self::Energy(value), length + 1))
+                    }
+                    0x4b => {
+                        let (value, length) = read_u24(remaining)?;
+                        Ok((Self::Gas(value), length + 1))
+                    }
+                    $( $object_id => {
+                        let (value, length) = $reader(remaining)?;
+                        Ok((Self::$name(value), length + 1))
+                    } )*
+                    object_id => Err(DecodeError::InvalidProperty(object_id)),
+                }
+            }
+        }
+    };
+}
+
+generate_element![
+    { 0x00, PacketId, u8, read_u8, "packet ID", "" },
+    { 0x51, Acceleration, u16, read_u16, "acceleration", "m/s²" },
+    { 0x01, Battery, u8, read_u8, "battery", "%"},
+    { 0x12, Co2, u16, read_u16, "CO2", "ppm"},
```

Comment:
what order are these in? Is that how they're listed in the spec?

Reply:
Yes, this is the order from the spec.